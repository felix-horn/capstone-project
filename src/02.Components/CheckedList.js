import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import ListItem from '../01.UI-Elements/ListItem'

CheckedList.propTypes = {
  list: PropTypes.object.isRequired,
  listAllocation: PropTypes.array.isRequired,
  handleInputChange: PropTypes.func.isRequired,
  toggleIsChecked: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired,
  addListItemOnEnter: PropTypes.func.isRequired,
  checked: PropTypes.bool,
}

export default function CheckedList({
  list,
  listAllocation,
  handleInputChange,
  toggleIsChecked,
  deleteListItem,
  addListItemOnEnter,
  checked,
}) {
  return (
    <>
      {listAllocation.length > 0 && (
        <ListStyled checked={checked}>
          {listAllocation.map((id) => {
            const { title, isChecked } = list.byId[id]
            return (
              <ListItem
                key={id}
                id={id}
                title={title}
                isChecked={isChecked}
                onInputChange={handleInputChange}
                onToggleCheckbox={() => toggleIsChecked(id)}
                onDelete={() => deleteListItem(id)}
                onEnter={() => addListItemOnEnter(id)}
              />
            )
          })}
        </ListStyled>
      )}
    </>
  )
}

const ListStyled = styled.div`
  margin: 3px 2px 10px 26px;
  padding-top: ${(props) => (props.checked ? '10px' : 'none')};
  border-top: ${(props) => (props.checked ? '1px solid #dadcdf' : 'none')};
  display: grid;
`
