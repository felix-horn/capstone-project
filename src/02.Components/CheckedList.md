```js padded
<List 
    listAllocation={["a", "b"]}
    list={ {       byId: {
            "a": { id: "a", title: "Koks", isChecked: false },
            "b": { id: "b", title: "Nutten", isChecked: false},
        }
    }}
/>

<List 
    listAllocation={["c", "d"]}
    list={ {       byId: {
            "c": { id: "c", title: "GummibÃ¤rchen", isChecked: true },
            "d": { id: "d", title: "Pumuckl-Torte", isChecked: true},
        }
    }}
/>
```