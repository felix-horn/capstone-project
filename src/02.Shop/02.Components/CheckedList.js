import PropTypes from 'prop-types'
import styled from 'styled-components/macro'

import { getCheckedItemIds } from '../../services/database.services'
import ListItem from '../01.UI-Elements/ListItem'

CheckedList.propTypes = {
  shopId: PropTypes.string.isRequired,
  database: PropTypes.object.isRequired,
  changeTitle: PropTypes.func.isRequired,
  toggleIsChecked: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired,
  className: PropTypes.func,
}

export default function CheckedList({
  shopId,
  database,
  changeTitle,
  toggleIsChecked,
  deleteListItem,
  className,
}) {
  const checkedIds = getCheckedItemIds(database, shopId)

  return (
    <>
      {checkedIds.length > 0 && (
        <ListLayout className={className}>
          {checkedIds.map((id) => (
            <ListItem
              key={id}
              itemId={id}
              shopId={shopId}
              database={database}
              title={database.items.byId[id].title}
              changeTitle={(fieldValue) => changeTitle(id, fieldValue)}
              toggleCheckbox={() => toggleIsChecked(id)}
              onDelete={() => deleteListItem(id)}
            />
          ))}
        </ListLayout>
      )}
    </>
  )
}

const ListLayout = styled.div`
  border-top: var(--border);
  padding-top: 10px;
  padding-left: 25px;
  display: grid;
`
