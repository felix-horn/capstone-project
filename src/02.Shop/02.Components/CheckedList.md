```js padded
<CheckedList
  shopId={'x'}
  database={{
    shops: {
      allIds: ['x'],
      byId: {
        x: {
          id: 'x',
          title: 'Shop X',
          items: ['a', 'b'],
        },
      },
    },
    items: {
      allIds: ['a', 'b'],
      byId: {
        a: {
          id: 'a',
          title: 'Item A',
          isChecked: true,
        },
        b: {
          id: 'b',
          title: 'Item B',
          isChecked: true,
        },
      },
    },
  }}
  changeTitle={() => {}}
  toggleIsChecked={() => {}}
  deleteListItem={() => {}}
/>
```
