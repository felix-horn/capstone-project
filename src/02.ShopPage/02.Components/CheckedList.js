import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import ListItem from '../01.UI-Elements/ListItem'

CheckedList.propTypes = {
  shopId: PropTypes.string.isRequired,
  database: PropTypes.object.isRequired,
  changeTitle: PropTypes.func.isRequired,
  toggleIsChecked: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired,
}

export default function CheckedList({
  shopId,
  database,
  changeTitle,
  toggleIsChecked,
  deleteListItem,
}) {
  const checkedIds = database.shops.byId[shopId].items.filter(
    (id) => database.items.byId[id].isChecked
  )
  return (
    <>
      {checkedIds.length > 0 && (
        <CheckedListStyled>
          {checkedIds.map((id) => (
            <ListItem
              key={id}
              itemId={id}
              shopId={shopId}
              database={database}
              title={database.items.byId[id].title}
              changeTitle={(fieldValue) => changeTitle(id, fieldValue)}
              toggleCheckbox={() => toggleIsChecked(id)}
              onDelete={() => deleteListItem(id)}
            />
          ))}
        </CheckedListStyled>
      )}
    </>
  )
}

const CheckedListStyled = styled.div`
  margin: 3px 2px 10px 0;
  padding-top: 10px;
  padding-left: 26px;
  border-top: var(--border);
  display: grid;
`
