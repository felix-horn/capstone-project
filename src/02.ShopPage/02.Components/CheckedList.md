```js padded
<CheckedList 
shopId={'x'} 
database={{
    shops: {
    allIds: ['x', 'y'],
    byId: {
      x: {
        id: 'x',
        title: 'Shop X',
        items: ['a', 'b'],
      },
      y: {
        id: 'y',
        title: 'Shop Y',
        items: ['c'],
      },
    },
  },
  items: {
    allIds: ['a', 'b', 'c'],
    byId: {
      a: {
        id: 'a',
        title: 'Item A',
        isChecked: true,
      },
      b: {
        id: 'b',
        title: 'Item B',
        isChecked: true,
      },
      c: {
        id: 'c',
        title: 'Item C',
        isChecked: true,
      },
    },
  },
/>
```