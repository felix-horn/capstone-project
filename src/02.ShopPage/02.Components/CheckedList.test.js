import { render } from '@testing-library/react'
import CheckedList from './CheckedList'

const testDatabase = {
  shops: {
    allIds: ['x', 'y', 'z'],
    byId: {
      x: {
        id: 'x',
        title: 'shopX',
        items: ['a', 'b'],
      },
      y: {
        id: 'y',
        title: 'shopY',
        items: ['c'],
      },
    },
  },
  items: {
    allIds: ['a', 'b', 'c'],
    byId: {
      a: {
        id: 'a',
        title: 'itemA',
        isChecked: false,
      },
      b: {
        id: 'b',
        title: 'itemB',
        isChecked: true,
      },
      c: {
        id: 'c',
        title: 'itemC',
        isChecked: false,
      },
    },
  },
}

const testProps = {
  changeTitle: jest.fn(),
  toggleIsChecked: jest.fn(),
  deleteListItem: jest.fn(),
  addListItem: jest.fn(),
  rearrangeListOrder: jest.fn(),
}

describe('CheckedList', () => {
  it('renders the list correctly', () => {
    const { container } = render(
      <CheckedList {...testProps} database={testDatabase} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  it('only renders allocated listItems', () => {
    const { queryByDisplayValue, getByDisplayValue } = render(
      <CheckedList {...testProps} database={testDatabase} />
    )
    const itemA = queryByDisplayValue(/itemA/i)
    const itemB = getByDisplayValue(/itemB/i)
    expect(itemA).not.toBeInTheDocument()
    expect(itemB).toBeInTheDocument()
  })
})
