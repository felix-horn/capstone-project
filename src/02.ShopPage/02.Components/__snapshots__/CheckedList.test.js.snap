// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckedList renders the list correctly 1`] = `
.c1 {
  width: 75vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 0.5;
}

.c2 {
  margin-left: 9px;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  border: none;
}

.c2:focus {
  outline: none;
}

.c3 {
  margin-left: auto;
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
  opacity: 0.4;
}

.c0 {
  margin: 3px 2px 10px 0;
  padding-top: 10px;
  padding-left: 26px;
  border-top: var(--border);
  display: grid;
}

<div
  class="c0"
>
  <label
    class="c1"
    data-testid="list-item"
  >
    <span
      aria-disabled="false"
      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root PrivateSwitchBase-checked-2 Mui-checked"
      data-testid="checkbox"
    >
      <span
        class="MuiIconButton-label"
      >
        <input
          checked=""
          class="PrivateSwitchBase-input-4"
          data-indeterminate="false"
          type="checkbox"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </span>
    <input
      class="c2"
      data-testid="title-list-item"
      value="itemB"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root c3"
      data-testid="delete-list-item"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </label>
</div>
`;
